src/storage/src/sink/kafka.rs

```diff
@@ -879,19 +907,43 @@ impl KafkaSinkState {
 
             if min_frontier > self.latest_progress_ts {
                 // record the write frontier in the progress topic.
-                if let Some(progress_state) = self.sink_state.unwrap_running() {
-                    self.retry_on_txn_error(|p| p.begin_transaction()).await;
+                let () = match self
+                    .producer
+                    .retry_on_txn_error(|p| p.begin_transaction())
+                    .await
+                {
+                    Ok(()) => (),
```

Comment:
This pattern gets repeated a ton in this branch... how would you feel about a replace with an `if let Err(err)`, or factoring it out to a helper?

Reply:
(I appreciate this was introduced to deal with my previous comment... 😅)

Reply:
I'm not a big fan of `if let Err ...` for future-proofing reasons and I think a helper basically puts us back to where we were before with the healthchecker being passed through.  I can do this though (with some futures combinators pre-await) -- but it requires adding an extra two lines for type annotations / because the NeverType isn't stabilized yet
```
               let () = s
                    .producer
                    .retry_on_txn_error(|p| p.init_transactions())
                    .or_else(|err| async {
                        Healthchecker::report_stall_and_panic(healthchecker.as_mut(), err).await
                    })
                    .await
                    .map_err(|()|())
                    .expect("unreachable: can remove once ! stabilizes");
```

It is a little junk (but junk that we should soon (tm) be able to get rid of!) -- but I think it's readable and more future proof than what's currently there!

What do you think?

Reply:
Mmm, I see what you're after there, but still feels a bit heavy to me.

Not worth burning too many cycles on probably, so happy to leave alone for now; thanks for considering!